Update Loadbalancer when there is a change in endpoints of a service of type=LoadBalancer

spradeepv opened the pull request:
**What this PR does / why we need it**:
 This PR compares the endpoints of a old service object with a new service object, inside a watcher in servicecontroller.go. If there is a difference in the count, UpdateLoadBalancer() function of the provider's loadbalancer implementation is invoked for a service of type=LoadBalancer.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #44203

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

spradeepv committed:
#44203

Fix for invoking UpdateLoadBalancer() for a service with type=LoadBalancer when pods are added/deleted

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44840)
<!-- Reviewable:end -->

spradeepv commented:
Signed CLA

xiangpengzhao commented:
@spradeepv Could you please amend your PR title and commit message to make them meaningful?

MrHohn commented:
/release-note-none

@k8s-bot ok to test

cc @freehan @nicksardo

MrHohn commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -125,16 +127,33 @@ func New(
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
+				oldEndpoints := s.endpoints[oldSvc.ObjectMeta.Name]
+				newEndpoints := s.getNumberOfEndpoints(oldSvc)
```
I don't think comparing number of endpoints is enough. What if one endpoint is removed while another is added?

sudhakso replied:
We may not want to check endpoints when the service is being created first time. Services that gets updated less frequently should see no additional overhead.

sudhakso replied:
@MrHohn We need this implementation to handle updates for services of type=loadbalancer. And that's a choice that each Cloud Provider would have to make (that should be made configurable), if it uses 'nodeIPs' versus 'podIPs' as the LB members for such services.

I think it is reasonable to implement a check to decide whether the implementation uses NodeIp versus PodIp, and design the corse of this implementation.

spradeepv replied:
@MrHohn I have addressed your comment

MrHohn reviewed (commented):
There is currently no requirement on any other cloud providers to call UpdateLoadBalancer when endpoints are updated. I could imagine in future we may want endpoints changes to trigger some sort of LB updates as well. Though this might be very much diverged in each cloud provider, I'm not sure if UpdateLoadBalancer() is the right entry for this functionality.

For now, this seems to introduce unnecessary function calls to all other cloud providers.

sudhakso commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -125,16 +127,33 @@ func New(
 			UpdateFunc: func(old, cur interface{}) {
+				endpointsChanged := false
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
```
Can these checks be includes as part of 'needsUpdate' function? That way we consolidate the decision making through that interface.

spradeepv replied:
Done

sudhakso commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -125,16 +127,33 @@ func New(
+				}
 				if ok1 && ok2 && s.needsUpdate(oldSvc, curSvc) {
 					s.enqueueService(cur)
+				} else if wantsLoadBalancer(curSvc) && endpointsChanged {
```
Can we move the calling part of UpdateLoadBalancer into createLoadBalancerIfNeeded(..) function? That would consolidate code that does action at one place.

spradeepv replied:
Done

sudhakso commented on pkg/controller/service/servicecontroller.go:
```diff
@@ -125,16 +127,33 @@ func New(
+						s.eventRecorder.Event(curSvc,
+							api.EventTypeNormal, 
+							"UpdatedLoadBalancer", 
+							"Updated load balancer with new Endpoints")
```
Can we handle event description in more generic way to handle 'add' and 'delete' members? This gives me an impression that it is considering addition of new endpoints only.

spradeepv replied:
Done

spradeepv committed:
Merge branch 'master' into master

spradeepv committed:
Update servicecontroller.go

Addressed comments. Now the map has service name as key and all the endpoints as values, which is used to compare and update the load balancer.

spradeepv committed:
Update servicecontroller.go

Removed old code

spradeepv committed:
Update servicecontroller.go

Fixed build failures

spradeepv commented:
Closing this pull request for now

andelwu commented:
This is helpful.
